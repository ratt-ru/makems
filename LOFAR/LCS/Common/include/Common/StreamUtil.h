//#  StreamUtil.h: useful stream manipulation methods.
//#
//#  Copyright (C) 2002-2003
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id: StreamUtil.h 9368 2006-10-03 14:02:18Z loose $

#ifndef LOFAR_COMMON_STREAMUTIL_H
#define LOFAR_COMMON_STREAMUTIL_H

// \file
// Useful stream manipulation methods.

//# Includes
#include <Common/LofarTypes.h>
#include <Common/lofar_string.h>
#include <Common/lofar_vector.h>
#include <Common/lofar_iosfwd.h>

namespace LOFAR
{
  // Handles indentation of text lines. Every time an Indent object is
  // constructed, the static member \c level is incremented by one, and every
  // time an Indent object is destructed \c level is decremented by one. To
  // increment the amount of indentation you simply create an Indent
  // object. When this object goes out of scope, the amount of indentation is
  // automagically decremented.
  class Indent {
  public:
    // Constructor. Increments indentation level.
    Indent() { lvl++; }
    // Destructor. Decrements indentation level.
    ~Indent() { lvl--; }
    // Return the amount of indentation.
    static uint level() { return lvl; }
    // Return the token to be printed per indentation level
    static const string& token() { return tok; }
  private:
    // Indentation level.
    static uint lvl;
    // Token to be printed per indentation level.
    static const string tok;
  };


  // Write a vector to an ostream with a given separator, prefix and postfix.
  template<class T>
  void writeVector (std::ostream& os, const std::vector<T>& vec,
		    const char* separator=",",
		    const char* prefix="[", const char* postfix="]")
  {
    os << prefix;
    for (uint i = 0; i < vec.size(); i++) {
      if (i > 0) os << separator;
      os << vec[i];
    }
    os << postfix;
  }


  // Print an indentation that depends on the number of Indent objects
  // currently in existence.
  inline ostream& indent(ostream& os) 
  {
    for (uint i = 0; i < Indent::level(); ++i) {
      os << Indent::token();
    }
    return os;
  }


  // Print the contents of a vector enclosed in square brackets, using a comma
  // as separator.
  // \note operator<<() must be defined for type \c T.
  template<typename T>
  inline ostream& operator<<(ostream& os, const vector<T>& v)
  {
    writeVector(os, v, ",", "[", "]");
    return os;
  }


} // namespace LOFAR

#endif
