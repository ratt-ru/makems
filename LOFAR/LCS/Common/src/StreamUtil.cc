//#  StreamUtil.cc: implementation of the stream utilities.
//#
//#  Copyright (C) 2002-2005
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id: StreamUtil.cc 9368 2006-10-03 14:02:18Z loose $

//# Always #include <lofar_config.h> first!
#include <lofar_config.h>

#include <Common/StreamUtil.h>

namespace LOFAR
{
  // Initialize static variables.
  uint Indent::lvl = 0;
  const string Indent::tok = ". ";

} // namespace LOFAR
