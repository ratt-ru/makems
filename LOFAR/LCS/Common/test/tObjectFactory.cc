//#  tObjectFactory.cc: Test program for the ObjectFactory class.
//#
//#  Copyright (C) 2006
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id: tObjectFactory.cc 9677 2006-11-22 21:18:08Z loose $

//# Always #include <lofar_config.h> first!
#include <lofar_config.h>

//# This program can only be compiled if Boost is available.
#ifdef HAVE_BOOST

//# Includes
#include "Shapes.h"
#include "Coordinates.h"
#include <Common/StreamUtil.h>
#include <memory>

using namespace LOFAR;
using namespace std;

void doShapes()
{
  cout << "Registered classes: "
       << ShapeFactory::instance().registeredClassIds()
       << endl;
  {
    auto_ptr<Shape> shape(ShapeFactory::instance().create("Rectangle"));
    cout << *shape << endl;
  }
  {
    auto_ptr<Shape> shape(ShapeFactory::instance().create("Square"));
    cout << *shape << endl;
  }
  {
    auto_ptr<Shape> shape(ShapeFactory::instance().create("Ellipse"));
    cout << *shape << endl;
  }
  {
    auto_ptr<Shape> shape(ShapeFactory::instance().create("Circle"));
    cout << *shape << endl;
  }
}

void doCoordinates()
{
  cout << "Registered classes: "
       << CoordinateFactory::instance().registeredClassIds()
       << endl;
  {
    auto_ptr<Coordinate> coord(CoordinateFactory::instance().
      create("Cartesian", 2.4, 3.6, 4.8));
    cout << *coord << endl;
  }
  {
    auto_ptr<Coordinate> coord(CoordinateFactory::instance().
      create("Cylindrical", 0.6, 1.2, 2.4));
    cout << *coord << endl;
  }
  {
    auto_ptr<Coordinate> coord(CoordinateFactory::instance().
      create("Spherical", 0.3, 0.15, 9.6));
    cout << *coord << endl;
  }
}

int main()
{
  cout << endl << "** Shapes **" << endl;
  doShapes();
  cout << endl << "** Coordinates **" << endl;
  doCoordinates();
  cout << endl;
  return 0;
}

#else

int main()
{
  // Test skipped.
  return 3;
}

#endif
