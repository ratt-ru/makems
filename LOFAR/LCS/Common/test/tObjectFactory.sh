#!/bin/sh
./runctest.sh -stdout tObjectFactory 2>&1 > tObjectFactory.log
