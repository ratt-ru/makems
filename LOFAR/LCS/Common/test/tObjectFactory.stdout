
** Shapes **
Registered classes: [Circle,Ellipse,Rectangle,Square]
Rectangle
Square
Ellipse
Circle

** Coordinates **
Registered classes: [Cartesian,Cylindrical,Spherical]
Cartesian: (x, y, z) = (2.4, 3.6, 4.8)
Cylindrical: (r, theta, z) = (0.6, 1.2, 2.4)
Spherical: (r, theta, phi) = (0.3, 0.15, 9.6)

